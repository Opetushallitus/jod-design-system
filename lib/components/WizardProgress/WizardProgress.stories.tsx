import type { Meta, StoryObj } from '@storybook/react';

import { WizardProgress } from './WizardProgress';

const meta = {
  title: 'Wizard/WizardProgress',
  component: WizardProgress,
  tags: ['autodocs'],
} satisfies Meta<typeof WizardProgress>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  decorators: [
    (Story) => (
      <div lang="fi">
        <button className="ds:sr-only">Interaktiivinen elementti</button>
        <Story />
      </div>
    ),
  ],
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=542-6678',
    },
    docs: {
      description: {
        story: 'This is a wizard progress component.',
      },
    },
  },
  args: {
    steps: 5,
    currentStep: 2,
    stepText: 'Vaihe',
    completedText: 'Valmis',
    currentText: 'Nykyinen',
    labelText: 'Eteneminen',
  },
};
