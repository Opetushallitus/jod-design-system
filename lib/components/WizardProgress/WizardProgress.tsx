import { MdCheck } from 'react-icons/md';

export interface WizardProgressProps {
  /** The total number of steps in the wizard. */
  steps: number;
  /** The current step in the wizard. */
  currentStep: number;
  /** The text for screenreader when a step is completed. */
  completedText: string;
  /** The text for screenreader when a step is the current step. */
  currentText: string;
  /** The text for describing meaning for screenreader for every step */
  stepText: string;
  /** The text for telling screenreader what the component is all about */
  labelText: string;
}

const Step = ({ step, stepText }: { step: number; stepText: string }) => {
  return (
    <>
      <span className="ds:sr-only">{`${stepText}: `}</span>
      <span className="ds:select-none">{step}</span>
    </>
  );
};

const CompletedStep = ({ text, step, stepText }: { text: string; step: number; stepText: string }) => {
  return (
    <>
      <span className="ds:sr-only">{`${stepText}: ${step}, ${text}.`}</span>
      <MdCheck role="presentation" size={24} />
    </>
  );
};

const CurrentStep = ({ text, step, stepText }: { text: string; step: number; stepText: string }) => {
  return (
    <>
      <Step step={step} stepText={stepText} />
      <span className="ds:sr-only">{`, ${text}.`}</span>
    </>
  );
};

/** A component that displays the progress of a wizard. */
export const WizardProgress = ({
  steps,
  currentStep,
  completedText,
  currentText,
  stepText,
  labelText,
}: WizardProgressProps) => {
  const renderStep = (step: number) => {
    if (step < currentStep) {
      return <CompletedStep text={completedText} step={step} stepText={stepText} />;
    } else if (step === currentStep) {
      return <CurrentStep text={currentText} step={step} stepText={stepText} />;
    } else {
      return <Step step={step} stepText={stepText} />;
    }
  };

  return (
    <ol className="ds:flex ds:gap-4 ds:text-card-label-lg ds:text-primary-gray" aria-label={labelText}>
      {Array.from({ length: steps }, (_, index) => (
        <li
          key={index + 1}
          className={`ds:flex ds:min-h-7 ds:min-w-7 ds:items-center ds:justify-center ds:rounded-full ${
            index + 1 === currentStep ? 'ds:bg-accent ds:text-white' : 'ds:bg-bg-gray-2'
          } ds:sm:min-h-8 ds:sm:min-w-8`}
          aria-current={index + 1 === currentStep}
        >
          {renderStep(index + 1)}
        </li>
      ))}
    </ol>
  );
};
