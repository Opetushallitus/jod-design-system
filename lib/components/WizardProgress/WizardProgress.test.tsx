import { render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import { WizardProgress } from './WizardProgress';

describe('WizardProgress', () => {
  const steps = 5;
  const currentStep = 3;
  const completedText = 'Completed';
  const currentText = 'Current';

  it('renders correctly', () => {
    const { container } = render(
      <WizardProgress
        labelText="Progress"
        stepText="Step"
        steps={steps}
        currentStep={currentStep}
        completedText={completedText}
        currentText={currentText}
        dataTestId="wiz"
      />,
    );
    expect(container.querySelector('[data-testid="wiz"]')).toBeInTheDocument();
    expect(container.querySelector('[data-testid="wiz-step-3"]')).toBeInTheDocument();
    // avoid snapshot churn due to attributes
  });
});
