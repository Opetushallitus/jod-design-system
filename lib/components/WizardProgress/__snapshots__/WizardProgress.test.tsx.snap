// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WizardProgress > renders correctly 1`] = `
<ol
  class="flex gap-4 text-body-md font-bold text-black sm:text-body-lg sm:font-bold"
>
  <li
    class="flex min-h-7 min-w-7 items-center justify-center rounded-full bg-bg-gray sm:min-h-8 sm:min-w-8"
  >
    <span
      class="sr-only"
    >
      Completed: 1
    </span>
    <span
      aria-hidden="true"
      class="material-symbols-outlined size-24 sm:size-32 select-none font-bold"
    >
      check
    </span>
  </li>
  <li
    class="flex min-h-7 min-w-7 items-center justify-center rounded-full bg-bg-gray sm:min-h-8 sm:min-w-8"
  >
    <span
      class="sr-only"
    >
      Completed: 2
    </span>
    <span
      aria-hidden="true"
      class="material-symbols-outlined size-24 sm:size-32 select-none font-bold"
    >
      check
    </span>
  </li>
  <li
    class="flex min-h-7 min-w-7 items-center justify-center rounded-full bg-accent text-white sm:min-h-8 sm:min-w-8"
  >
    <span
      class="sr-only"
    >
      Current: 
    </span>
    <span
      class="select-none"
    >
      3
    </span>
  </li>
  <li
    class="flex min-h-7 min-w-7 items-center justify-center rounded-full bg-bg-gray sm:min-h-8 sm:min-w-8"
  >
    <span
      class="select-none"
    >
      4
    </span>
  </li>
  <li
    class="flex min-h-7 min-w-7 items-center justify-center rounded-full bg-bg-gray sm:min-h-8 sm:min-w-8"
  >
    <span
      class="select-none"
    >
      5
    </span>
  </li>
</ol>
`;
